lib/Index/Index.cpp

```diff
@@ -440,6 +464,270 @@ class IndexSwiftASTWalker : public SourceEntityWalker {
     return true;
   }
 
+  bool walkToTypeReprPre(TypeRepr *T) override {
+    if (Cancelled)
+      return false;
+    activateContainerDecls(T);
+    return true;
+  }
+
+  bool walkToPatternPre(Pattern *P) override {
+    if (Cancelled)
+      return false;
+    activateContainerDecls(P);
+    return true;
+  }
+
+  /// Identify container declarations. Any references associated with these
+  /// declarations will obtain the \c RelationContainedBy role.
+  ///
+  /// The use of \c PatternBindingDecl here ensures that variable declarations
+  /// will contain references to explicit types, initializers, custom
+  /// attributes, etc.
+  void beginBalancedASTOrderDeclVisit(Decl *decl) override {
+    if (auto bindingPattern = dyn_cast<PatternBindingDecl>(decl)) {
+      ContainerStackEntry entry{decl};
+      identifyContainerVarDecls(bindingPattern, entry);
+      containerStack.push_back(std::move(entry));
+    } else if (isa<AbstractFunctionDecl>(decl)) {
+      containerStack.push_back({decl, {decl}});
+    }
+  }
+
+  void endBalancedASTOrderDeclVisit(Decl *decl) override {
+    if (containerStack.empty())
+      return;
+    if (containerStack.back().visitedDecl == decl)
+      containerStack.pop_back();
+  }
+
+  /// Identifies the VarDecls within a PatternBindingDecl along with the types
+  /// and expressions to which each corresponds.
+  ///
+  /// Given the tuple binding 'let (a, b): (Int, String) = (intValue,
+  /// stringValue)' we can see that 'a' corresponds to 'Int' and 'intValue',
+  /// while 'b' corresponds to 'String' and 'stringValue'. By identifying these
+  /// relationships, we can pinpoint locations within the PatternBindingDecl
+  /// that the corresponding VarDecl should be marked as the current active
+  /// container. For example, when we walk to the TypeRepr for 'Int', we can
+  /// make 'a' the current active container, and likewise for 'String', 'b'. And
+  /// thus, when the walk reaches the point of creating the reference for 'Int'
+  /// it will be contained by 'a'.
+  ///
+  /// These locations are also identified for single named pattern bindings; in
+  /// which case, the VarDecl is activated for all types and expressions with
+  /// the pattern.
+  ///
+  /// Pattern bindings containing an AntPattern (i.e let _) are a special case,
+  /// as they have no VarDecl. Their types and expressions are associated with
+  /// the current active container, if any. Therefore, given such a pattern
+  /// declared within a function, the type and expression references will be
+  /// contained by the function. If there is no active container, the references
+  /// are not contained.
+  void identifyContainerVarDecls(const PatternBindingDecl *const bindingDecl,
+                                 ContainerStackEntry &entry) const {
```

Comment:
`ContainerStackEntry` could be the return for this function (possibly with a rename though). There's a few other functions that are similar as well, ie. `collectPatternVarDecls` could just return the map.